misoc/targets/kasli.py

```diff
@@ -116,90 +114,42 @@ def __init__(self, platform):
-        else:
-            cdr_clk_out = platform.request("si5324_clkout")
-        
-        self.cdr_clk_buf = Signal()
-        cdr_clk_div2 = Signal()
-
-        platform.add_period_constraint(cdr_clk_out, 8.)
-
-        self.specials += Instance("IBUFDS_GTE2",
-            i_CEB=0,
-            i_I=cdr_clk_out.p, i_IB=cdr_clk_out.n,
-            o_O=self.cdr_clk_buf,
-            o_ODIV2=cdr_clk_div2)
+        self.specials += Instance("BUFG", i_I=self.clk125_div2, o_O=self.cd_bootstrap.clk)
 
-        clk_sw_fsm = ClockSwitchFSM()
-        self.submodules += clk_sw_fsm
+        # used internally in case main clock output is 125MHz
+        self._clk125_mult = Signal()
 
-        mmcm_locked = Signal()
-        mmcm_fb_in = Signal()
-        mmcm_fb_out = Signal()
-        mmcm_sys = Signal()
-        mmcm_sys4x = Signal()
-        mmcm_sys4x_dqs = Signal()
+        self.clk_sw_fsm = ClockSwitchFSM()
+        self.submodules += self.clk_sw_fsm
 
         pll_clk200 = Signal()
+        pll_clk125 = Signal()
         pll_fb = Signal()
-        pll_locked = Signal()
+        self.pll_locked = Signal()
         self.specials += [
-            Instance("MMCME2_ADV",
-                p_CLKIN1_PERIOD=16.0,
-                i_CLKIN1=cdr_clk_div2,
-                i_CLKIN2=self.clk125_div2,
-
-                i_CLKINSEL=clk_sw_fsm.o_clk_sw,
-                i_RST=clk_sw_fsm.o_reset,
-
-                i_CLKFBIN=mmcm_fb_in,
-                o_CLKFBOUT=mmcm_fb_out,
-                o_LOCKED=mmcm_locked,
-
-                # VCO @ 1GHz with MULT=16
-                p_CLKFBOUT_MULT_F=16, p_DIVCLK_DIVIDE=1,
-
-                # ~125MHz
-                # kasli 1.1 seems to still be within timing on that
-                p_CLKOUT0_DIVIDE_F=8, p_CLKOUT0_PHASE=0.0, o_CLKOUT0=mmcm_sys,
-
-                # ~500MHz. Must be more than 400MHz as per DDR3 specs.
-                p_CLKOUT1_DIVIDE=2, p_CLKOUT1_PHASE=0.0, o_CLKOUT1=mmcm_sys4x,
-                p_CLKOUT2_DIVIDE=2, p_CLKOUT2_PHASE=90.0, o_CLKOUT2=mmcm_sys4x_dqs,
-            ),
             Instance("PLLE2_BASE",
                 p_CLKIN1_PERIOD=16.0,
                 i_CLKIN1=self.clk125_div2,
 
                 i_CLKFBIN=pll_fb,
                 o_CLKFBOUT=pll_fb,
-                o_LOCKED=pll_locked,
+                o_LOCKED=self.pll_locked,
 
                 # VCO @ 1GHz
                 p_CLKFBOUT_MULT=16, p_DIVCLK_DIVIDE=1,
 
                 # 200MHz for IDELAYCTRL
                 p_CLKOUT0_DIVIDE=5, p_CLKOUT0_PHASE=0.0, o_CLKOUT0=pll_clk200,
+                # 125MHz for non-div2 MMCM
+                p_CLKOUT1_DIVIDE=8, p_CLKOUT1_PHASE=0.0, o_CLKOUT1=pll_clk125
             ),
-            Instance("BUFG", i_I=mmcm_sys, o_O=self.cd_sys.clk),
-            Instance("BUFG", i_I=mmcm_sys4x, o_O=self.cd_sys4x.clk),
-            Instance("BUFG", i_I=mmcm_sys4x_dqs, o_O=self.cd_sys4x_dqs.clk),
+            Instance("BUFG", i_I=pll_clk125, o_O=self._clk125_mult),
```

Comment:
BUFG is necessary not because of the PLL to MMCM, but because of txout/si5324 output - and Vivado will complain that the clock sources for the MMCM outputs are not the same.